File: csharp/AoC2021/Day22/Day22/Cuboid.cs
namespace Day22;

public record Cuboid(bool IsOn, Range XRange, Range YRange, Range ZRange)
{
    public bool IsDisjoint(Cuboid other) =>
        XRange.IsDisjoint(other.XRange)
        || YRange.IsDisjoint(other.YRange)
        || ZRange.IsDisjoint(other.ZRange);

    public bool FitsWithin(Cuboid other) =>
        XRange.FitsWithin(other.XRange)
        && YRange.FitsWithin(other.YRange)
        && ZRange.FitsWithin(other.ZRange);

    public long Size() => XRange.Size() * YRange.Size() * ZRange.Size();

    /// <summary>
    /// Return a selection of cuboids that together contain all cubes that are in <paramref name="one"/>, but not in
    /// <paramref name="other"/>.
    /// </summary>
    /// <param name="one">The cuboid from which cubes are subtracted.</param>
    /// <param name="other">The cuboid that will be subtracted.</param>
    /// <returns>Cuboids containing the cubes that are in <paramref name="one"/>, but not <paramref name="other"/>.</returns>
    public static IEnumerable<Cuboid> operator -(Cuboid one, Cuboid other) =>
        one.XRange.Partition(other.XRange)
            .SelectMany(x => one
                .YRange
                .Partition(other.YRange)
                .Select(y => (x, y))
            )
            .SelectMany(ranges => one
                .ZRange
                .Partition(other.ZRange)
                .Select(z => (ranges.x, ranges.y, z))
            )
            .Select(ranges => new Cuboid(one.IsOn, ranges.x, ranges.y, ranges.z))
            .Where(cuboid => !cuboid.FitsWithin(other));

    public static Cuboid Parse(string s)
    {
        var i = FindIndices(s);
        return new Cuboid(
            s.StartsWith("on"),
            Range.Parse(s[i[0]..i[1]]),
            Range.Parse(s[i[2]..i[3]]),
            Range.Parse(s[i[4]..])
        );
    }

    private static int[] FindIndices(string s)
    {
        var i = new[] { 0, 0, 0, 0, 0 };

        // XRange.From
        i[0] = s.IndexOf("x=", StringComparison.Ordinal);
        if (i[0] == -1) throw new FormatException($"Invalid input: {s}");
        i[0] += 2;

        // XRange.To
        i[1] = s.IndexOf(",y=", i[0], StringComparison.Ordinal);
        if (i[1] == -1) throw new FormatException($"Invalid input: {s}");

        // YRange.From
        i[2] = i[1] + 3;

        // YRange.To
        i[3] = s.IndexOf(",z=", i[2], StringComparison.Ordinal);
        if (i[3] == -1) throw new FormatException($"Invalid input: {s}");

        // ZRange.From
        i[4] = i[3] + 3;

        return i;
    }
}

File: rust/aoc2021/src/day22/cuboid.rs
use std::fmt;
use std::ops::Sub;
use std::str::FromStr;

use rdcl_aoc_helpers::error::ParseError;
use rdcl_aoc_helpers::parse_error;

use crate::range::Range;

#[derive(Copy, Clone)]
pub struct Cuboid {
    pub is_on: bool,
    pub x_range: Range,
    pub y_range: Range,
    pub z_range: Range,
}

impl Cuboid {
    pub fn is_disjoint(&self, other: &Cuboid) -> bool {
        self.x_range.is_disjoint(&other.x_range)
            || self.y_range.is_disjoint(&other.y_range)
            || self.z_range.is_disjoint(&other.z_range)
    }

    pub fn fits_within(&self, other: &Cuboid) -> bool {
        self.x_range.fits_within(&other.x_range)
            && self.y_range.fits_within(&other.y_range)
            && self.z_range.fits_within(&other.z_range)
    }

    pub fn size(&self) -> usize {
        self.x_range.size() * self.y_range.size() * self.z_range.size()
    }
}

impl<'a> Sub<&'a Cuboid> for &'a Cuboid {
    type Output = Vec<Cuboid>;

    /// Return a selection of cuboids that together contain all points that are in the first cuboid,
    /// but not in the second.
    fn sub(self, other: &Cuboid) -> Self::Output {
        self.x_range
            .partition(&other.x_range)
            .iter()
            .flat_map(|x_range| {
                self.y_range
                    .partition(&other.y_range)
                    .iter()
                    .map(|y_range| (*x_range, *y_range))
                    .collect::<Vec<(Range, Range)>>()
            })
            .flat_map(|(x_range, y_range)| {
                self.z_range
                    .partition(&other.z_range)
                    .iter()
                    .map(|z_range| (x_range, y_range, *z_range))
                    .collect::<Vec<(Range, Range, Range)>>()
            })
            .map(|(x_range, y_range, z_range)| Cuboid {
                is_on: self.is_on,
                x_range,
                y_range,
                z_range,
            })
            .filter(|cuboid| !cuboid.fits_within(other))
            .collect()
    }
}

impl fmt::Debug for Cuboid {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{} x={:?},y={:?},z={:?}",
            if self.is_on { "on" } else { "off" },
            self.x_range,
            self.y_range,
            self.z_range,
        )
    }
}

impl FromStr for Cuboid {
    type Err = ParseError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let i = if let Some(v) = find_indices(s) {
            v
        } else {
            return Err(parse_error!("Invalid input string: {}", s));
        };

        Ok(Cuboid {
            is_on: s.starts_with("on"),
            x_range: s[i[0]..i[1]].parse()?,
            y_range: s[i[2]..i[3]].parse()?,
            z_range: s[i[4]..].parse()?,
        })
    }
}

fn find_indices(s: &str) -> Option<[usize; 5]> {
    let mut i = [0; 5];

    // x_range.from
    i[0] = 2 + s.find("x=")?;

    // x_range.to
    i[1] = i[0] + s[i[0]..].find(",y=")?;

    // y_range.from
    i[2] = i[1] + 3;

    // y_range.to
    i[3] = i[2] + s[i[2]..].find(",z=")?;

    // z_range.from
    i[4] = i[3] + 3;

    Some(i)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_fits_within() {
        let cuboid_a = Cuboid {
            is_on: true,
            x_range: Range::new(-5, 5),
            y_range: Range::new(-5, 5),
            z_range: Range::new(-5, 5),
        };
        let cuboid_b = Cuboid {
            is_on: true,
            x_range: Range::new(-10, 10),
            y_range: Range::new(-10, 10),
            z_range: Range::new(-10, 10),
        };

        assert!(cuboid_a.fits_within(&cuboid_a));
        assert!(cuboid_a.fits_within(&cuboid_b));
        assert!(cuboid_b.fits_within(&cuboid_b));
        assert!(!cuboid_b.fits_within(&cuboid_a));
    }

    #[test]
    fn test_size() {
        let cuboid = Cuboid {
            is_on: true,
            x_range: Range::new(-5, 5),
            y_range: Range::new(-5, 5),
            z_range: Range::new(-5, 5),
        };
        assert_eq!(cuboid.size(), 11 * 11 * 11);
    }

    #[test]
    fn test_from_str_1() {
        let cuboid = "on x=1..2,y=3..4,z=5..6".parse::<Cuboid>().unwrap();

        assert!(cuboid.is_on);
        assert_eq!(cuboid.x_range.from, 1);
        assert_eq!(cuboid.x_range.to, 2);
        assert_eq!(cuboid.y_range.from, 3);
        assert_eq!(cuboid.y_range.to, 4);
        assert_eq!(cuboid.z_range.from, 5);
        assert_eq!(cuboid.z_range.to, 6);
    }

    #[test]
    fn test_from_str_2() {
        let cuboid = "off x=1..2,y=3..4,z=5..6".parse::<Cuboid>().unwrap();

        assert!(!cuboid.is_on);
        assert_eq!(cuboid.x_range.from, 1);
        assert_eq!(cuboid.x_range.to, 2);
        assert_eq!(cuboid.y_range.from, 3);
        assert_eq!(cuboid.y_range.to, 4);
        assert_eq!(cuboid.z_range.from, 5);
        assert_eq!(cuboid.z_range.to, 6);
    }

    #[test]
    fn test_from_str_3() {
        let result = "invalid".parse::<Cuboid>();
        assert!(result.is_err());
    }
}
